File: src/main/java/brightspark/landmanager/message/MessageHomeToggle.java
package brightspark.landmanager.message;

import brightspark.landmanager.LMConfig;
import brightspark.landmanager.LandManager;
import brightspark.landmanager.data.areas.Area;
import brightspark.landmanager.data.areas.CapabilityAreas;
import brightspark.landmanager.util.HomeGuiToggleType;
import brightspark.landmanager.util.Utils;
import io.netty.buffer.ByteBuf;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

public class MessageHomeToggle implements IMessage
{
	private BlockPos pos;
	private HomeGuiToggleType type;

	public MessageHomeToggle() {}

	public MessageHomeToggle(BlockPos pos, HomeGuiToggleType type)
	{
		this.pos = pos;
		this.type = type;
	}

	@Override
	public void fromBytes(ByteBuf buf)
	{
		pos = BlockPos.fromLong(buf.readLong());
		type = HomeGuiToggleType.values()[buf.readByte()];
	}

	@Override
	public void toBytes(ByteBuf buf)
	{
		buf.writeLong(pos.toLong());
		buf.writeByte(type.ordinal());
	}

	public static class Handler implements IMessageHandler<MessageHomeToggle, MessageHomeToggleReply>
	{
		@Override
		public MessageHomeToggleReply onMessage(MessageHomeToggle message, MessageContext ctx)
		{
			EntityPlayerMP player = ctx.getServerHandler().player;
			CapabilityAreas cap = player.world.getCapability(LandManager.CAPABILITY_AREAS, null);
			if(cap == null)
				return null;
			Area area = cap.intersectingArea(message.pos);
			boolean playerIsOp = Utils.isOp(player.world.getMinecraftServer(), player);
			if(area == null || (!area.isOwner(player.getUniqueID()) && !playerIsOp))
				return null;
			HomeGuiToggleType type = message.type;
			switch(type)
			{
				case INTERACTIONS:
					if(!playerIsOp && !LMConfig.permissions.interactions)
						return null;
					area.toggleInteractions();
					cap.dataChanged();
					return new MessageHomeToggleReply(type, area.canInteract());
				case PASSIVE_SPAWNS:
					if(!playerIsOp && !LMConfig.permissions.passiveSpawning)
						return null;
					area.togglePassiveSpawning();
					cap.dataChanged();
					return new MessageHomeToggleReply(type, area.canPassiveSpawn());
				case HOSTILE_SPAWNS:
					if(!playerIsOp && !LMConfig.permissions.hostileSpawning)
						return null;
					area.toggleHostileSpawning();
					cap.dataChanged();
					return new MessageHomeToggleReply(type, area.canHostileSpawn());
				case EXPLOSIONS:
					if(!playerIsOp && !LMConfig.permissions.explosions)
						return null;
					area.toggleExplosions();
					cap.dataChanged();
					return new MessageHomeToggleReply(type, area.canExplosionsCauseDamage());
			}
			return null;
		}
	}
}

File: src/main/kotlin/brightspark/landmanager/message/MessageHomeToggle.kt
package brightspark.landmanager.message

import brightspark.landmanager.LMConfig
import brightspark.landmanager.LandManager
import brightspark.landmanager.data.areas.Area
import brightspark.landmanager.data.areas.AreaUpdateType
import brightspark.landmanager.data.areas.AreasCapability
import brightspark.landmanager.util.*
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.network.PacketBuffer
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.network.NetworkEvent
import java.util.function.Supplier

class MessageHomeToggle : Message {
	private lateinit var pos: BlockPos
	private lateinit var type: HomeGuiToggleType

	@Suppress("unused")
	constructor()

	constructor(pos: BlockPos, type: HomeGuiToggleType) {
		this.pos = pos
		this.type = type
	}

	override fun encode(buffer: PacketBuffer): Unit = buffer.run {
		writeBlockPos(pos)
		writeEnumValue(type)
	}

	override fun decode(buffer: PacketBuffer): Unit = buffer.run {
		pos = readBlockPos()
		type = readEnumValue()
	}

	override fun consume(context: Supplier<NetworkEvent.Context>) {
		context.get().enqueueWork {
			val player = context.get().sender ?: return@enqueueWork
			val world = player.world
			val cap = world.areasCap
			val area = cap.intersectingArea(pos) ?: return@enqueueWork
			if (!player.canEditArea(area))
				return@enqueueWork
			val playerIsOp = player.isOp()
			when (type) {
				HomeGuiToggleType.INTERACTIONS -> handleType(cap, player, area, playerIsOp, LMConfig.interactions, { it.toggleInteractions() }, { it.interactions })
				HomeGuiToggleType.PASSIVES -> handleType(cap, player, area, playerIsOp, LMConfig.passiveSpawning, { it.togglePassiveSpawning() }, { it.canPassiveSpawn })
				HomeGuiToggleType.HOSTILES -> handleType(cap, player, area, playerIsOp, LMConfig.hostileSpawning, { it.toggleHostileSpawning() }, { it.canHostileSpawn })
				HomeGuiToggleType.EXPLOSIONS -> handleType(cap, player, area, playerIsOp, LMConfig.explosions, { it.toggleExplosions() }, { it.explosions })
				else -> Unit
			}
		}
	}

	private fun handleType(
		cap: AreasCapability,
		player: ServerPlayerEntity,
		area: Area,
		playerIsOp: Boolean,
		config: Boolean,
		toggle: (Area) -> Unit,
		stateGetter: (Area) -> Boolean
	) {
		if (!playerIsOp && config)
			return
		toggle(area)
		cap.dataChanged(area, AreaUpdateType.CHANGE)
		LandManager.NETWORK.sendToPlayer(MessageHomeToggleReply(type, stateGetter(area)), player)
	}
}
